src/ir_Vestel.cpp

```diff
@@ -0,0 +1,436 @@
+// Set AC's wake up time. Takes time in minute.
+void IRVestelAC::setWakeupTime(const uint16_t minutes) {
+  remote_state.t_turnOnHour = minutes / 60;
+  // Register is 3 bit and can only hold minute/10
+  remote_state.t_turnOnMinute =(minutes%60) / 10;
+  remote_state.t_on_active = 1;  // Automatic activation
+}
+
+// Set AC's turn off time. Takes time in minute.
+void IRVestelAC::setTurnOffTime(const uint16_t minutes) {
+  remote_state.t_turnOffHour = minutes / 60;
+  // Register is 3 bit and can only hold minute/10
+  remote_state.t_turnOffMinute = (minutes%60) / 10;
+  remote_state.t_off_active = 1;  // Automatic activation
+}
+
+// Set the Sleep state of the A/C.
+void IRVestelAC::setSleep(const bool state) {
+  remote_state.turbo_sleep_normal = state ? kVestelACSleep : kVestelACNormal;
+}
+
+// Return the Sleep state of the A/C.
+bool IRVestelAC::getSleep() {
+  return remote_state.turbo_sleep_normal == kVestelACSleep;
+}
+
+// Set the Turbo state of the A/C.
+void IRVestelAC::setTurbo(const bool state) {
+  remote_state.turbo_sleep_normal = state ? kVestelACTurbo : kVestelACNormal;
+}
+
+// Return the Turbo state of the A/C.
+bool IRVestelAC::getTurbo() {
+  return remote_state.turbo_sleep_normal == kVestelACTurbo;
+}
+
+// Set the Ion state of the A/C.
+void IRVestelAC::setIon(const bool state) {
+  remote_state.ion = state ? kVestelACIon : 0;
+}
+
+// Return the Ion state of the A/C.
+bool IRVestelAC::getIon() { return remote_state.ion == kVestelACIon; }
+
+// Set the Swing Roaming state of the A/C.
+void IRVestelAC::setSwing(const bool state) {
+  remote_state.swing = state ? kVestelACSwing : 0xF;
+}
+
+// Return the Swing Roaming state of the A/C.
+bool IRVestelAC::getSwing() { return remote_state.swing == kVestelACSwing; }
+
+// Calculate the checksum for a given array.
+// Args:
+//   state:  The state to calculate the checksum over.
+// Returns:
+//   The 8 bit checksum value.
+uint8_t IRVestelAC::calcChecksum(const uint64_t state) {
+  // Just counts the set bits +1 on stream and take inverse after mask
+  uint8_t sum = 0;
+  uint64_t temp_state = state & kVestelACCRCMask;
+  for (; temp_state; temp_state >>= 1)
+    if (temp_state & 1) sum++;
+  sum += 2;
+  sum = 0xff - sum;
+  return sum;
+}
+
+// Verify the checksum is valid for a given state.
+// Args:
+//   state:  The state to verify the checksum of.
+// Returns:
+//   A boolean.
+bool IRVestelAC::validChecksum(const uint64_t state) {
+  return (((state >> 12) & 0xFF) == calcChecksum(state));
+}
+
+// Calculate & set the checksum for the current internal state of the remote.
+void IRVestelAC::checksum() {
+  // Stored the checksum value in the last byte.
+  remote_state.CRC = calcChecksum(remote_state.rawCode);
+}
+
+// Convert the internal state into a human readable string.
+#ifdef ARDUINO
+String IRVestelAC::toString() {
+  String result = "";
+#else
+std::string IRVestelAC::toString() {
+  std::string result = "";
+#endif  // ARDUINO
+  if (remote_state.power == 0x00) {
```

Comment:
Is this correct? I don't see anything that ever sets this to `0`. only `0xF` or `0xC`

Reply:
Yes. This is main difference to identify normal command stack or time stack.

If remote_state.power byte is not set, that means we are decoding time command.

I don't add any reset for that bits since not planning to implement timing functions.
I reset it using t_not_used.
But using power=0x00 might be better idea.

Reply:
My vote is to use `power` rather than `not_used` as clearly it is used.

Can you give me a code for a timer message so I can build some unit tests around it and fix toString() to reflect that.

Reply:
I just add new remote state for time.. So switching is not needed anymore.

Here is the some codes. "Mesg Desc" shows the time command function :)

```
Timestamp : 000008.851
Encoding  : VESTEL_AC
Code      : 2D6570B8EE201 (56 bits)
Mesg Desc.: Time Command - Time : 5:45, Turn On: 14:00, Turn Off: 23:00
Library   : v2.5.4

Raw Timing[115]:
   +  3022, -  9080,    +   546, -  1536,    +   526, -   466,    +   526, -   492, 
   +   526, -   468,    +   526, -   492,    +   524, -   468,    +   524, -   494, 
   +   524, -   504,    +   540, -   492,    +   524, -  1538,    +   526, -   468, 
   +   524, -   492,    +   526, -   466,    +   552, -  1536,    +   526, -  1536, 
   +   526, -  1570,    +   542, -   492,    +   524, -  1538,    +   550, -  1538, 
   +   524, -  1536,    +   526, -   494,    +   524, -   466,    +   526, -   468, 
   +   524, -  1574,    +   540, -  1536,    +   550, -  1536,    +   526, -   468, 
   +   550, -  1536,    +   526, -   492,    +   526, -   468,    +   524, -   492, 
   +   526, -   518,    +   526, -  1536,    +   552, -  1536,    +   550, -  1536, 
   +   526, -   494,    +   550, -  1538,    +   526, -   492,    +   524, -  1538, 
   +   526, -   504,    +   540, -   466,    +   526, -  1536,    +   526, -  1536, 
   +   526, -   468,    +   550, -  1538,    +   524, -   468,    +   524, -  1538, 
   +   550, -  1574,    +   540, -   468,    +   550, -  1538,    +   526, -   492, 
   +   524, -   468,    +   526, -   466,    +   526, -   468,    +   524, -   494, 
   +   524, -   468,    +   546

uint16_t rawData[115] = {3022, 9080,  546, 1536,  526, 466,  526, 492,  526, 468,  526, 492,  524, 468,  524, 494,  524, 504,  540, 492,  524, 1538,  526, 468,  524, 492,  526, 466,  552, 1536,  526, 1536,  526, 1570,  542, 492,  524, 1538,  550, 1538,  524, 1536,  526, 494,  524, 466,  526, 468,  524, 1574,  540, 1536,  550, 1536,  526, 468,  550, 1536,  526, 492,  526, 468,  524, 492,  526, 518,  526, 1536,  552, 1536,  550, 1536,  526, 494,  550, 1538,  526, 492,  524, 1538,  526, 504,  540, 466,  526, 1536,  526, 1536,  526, 468,  550, 1538,  524, 468,  524, 1538,  550, 1574,  540, 468,  550, 1538,  526, 492,  524, 468,  526, 466,  526, 468,  524, 494,  524, 468,  546};  // VESTEL_AC 2D6570B8EE201
uint64_t data = 0x2D6570B8EE201;
```

```
Timestamp : 000053.163
Encoding  : VESTEL_AC
Code      : 2EA50018F3201 (56 bits)
Mesg Desc.: Time Command - Time : 5:46, Timer Mode Off After 03:00
Library   : v2.5.4

Raw Timing[115]:
   +  3100, -  9076,    +   550, -  1534,    +   528, -   490,    +   528, -   464, 
   +   528, -   490,    +   528, -   464,    +   528, -   490,    +   528, -   466, 
   +   526, -   502,    +   542, -   464,    +   554, -  1534,    +   528, -   490, 
   +   528, -   464,    +   552, -  1534,    +   528, -  1534,    +   528, -   464, 
   +   528, -   526,    +   542, -  1534,    +   552, -  1534,    +   528, -  1536, 
   +   526, -  1534,    +   528, -   466,    +   528, -   464,    +   528, -   492, 
   +   526, -  1572,    +   542, -  1534,    +   528, -   490,    +   528, -   490, 
   +   526, -   466,    +   528, -   490,    +   528, -   464,    +   528, -   466, 
   +   526, -   502,    +   542, -   464,    +   528, -   490,    +   528, -   490, 
   +   528, -   466,    +   552, -  1534,    +   528, -   490,    +   528, -  1534, 
   +   528, -   526,    +   542, -   490,    +   554, -  1534,    +   528, -   466, 
   +   552, -  1534,    +   528, -   464,    +   552, -  1536,    +   552, -  1534, 
   +   528, -  1570,    +   544, -   490,    +   528, -  1534,    +   528, -   466, 
   +   526, -   490,    +   528, -   466,    +   528, -   490,    +   528, -   490, 
   +   526, -   466,    +   550

uint16_t rawData[115] = {3100, 9076,  550, 1534,  528, 490,  528, 464,  528, 490,  528, 464,  528, 490,  528, 466,  526, 502,  542, 464,  554, 1534,  528, 490,  528, 464,  552, 1534,  528, 1534,  528, 464,  528, 526,  542, 1534,  552, 1534,  528, 1536,  526, 1534,  528, 466,  528, 464,  528, 492,  526, 1572,  542, 1534,  528, 490,  528, 490,  526, 466,  528, 490,  528, 464,  528, 466,  526, 502,  542, 464,  528, 490,  528, 490,  528, 466,  552, 1534,  528, 490,  528, 1534,  528, 526,  542, 490,  554, 1534,  528, 466,  552, 1534,  528, 464,  552, 1536,  552, 1534,  528, 1570,  544, 490,  528, 1534,  528, 466,  526, 490,  528, 466,  528, 490,  528, 490,  526, 466,  550};  // VESTEL_AC 2EA50018F3201
uint64_t data = 0x2EA50018F3201;
```